### DISCUSSION TITLE
Packaging a GLPS application

### DISCUSSION THREAD

**Discussion by PatrikN-work** [eclipse-glsp/glsp#1010] [Q&A] [answered]
How can i package a GLSP application, i.e. the server and client, as a VSCode extension, ie. an .vsix file?

For example the https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/node-json-theia

**Comment by tortmayr** [comment]
Hi @PatrikN-work,

in the current state the project templates are no ready for packing/publishing.
The offical VS Code documentation contains a quite detailed [guide](https://code.visualstudio.com/api/working-with-extensions/publishing-extension) on how to publish extensions.
In general, the `vsce` CLI tool is used for creating a packed .vsix file and publishing to the VSCode market place.
In theory, you can package the extension as is with `vsce`, however, it is recommended do to bundle the extension content into a single file (https://code.visualstudio.com/api/working-with-extensions/bundling-extension). This reduces the overall extension size and removes not needed dependencies.

I have prepared a demo branch for the `node-json-vscode` project template that adapts the code so that the extension can be packaged efficiently: https://github.com/eclipse-glsp/glsp-examples/tree/packing-demo
The changes include:
- bundling of the main extension with webpack
- Adpating the code base to correctly use the bundled files
- Adding `vsce` as dependeny and providing a `package` script to create a vsix file.

If you checkout the branch and execute `yarn package` in the `glsp-client` (after building) a `*.vsix` file should be generated in the extension folder.

I have also opened a [discussion](https://github.com/eclipse-glsp/glsp/issues/1012) to improve the project-templates with regard to packaging for the upcoming 2.0 release.

BR, Tobias

**Comment by rsoika** [comment]
@PatrikN-work, you can also have a look in our [VS-Code Integration project](https://github.com/imixs/open-bpmn-vscode-integration). It may give you maybe some ideas and code snippets which you may reuse.

**Comment by PatrikN-work** [comment]
Thanks,

When is v2.0 out and what are the anticipated changes?

**Comment by tortmayr** [comment, marked as answer]
The 2.0.0 Release is schedules for end of June/Beginning of July and provides
a lot of new features, bugfixes and usability improvements.

 A complete Release Note with a detailled
change log is not yet available but here is a rough overview of the anticipated changes:
- Various bugfixes across all components
- Update to the latest sprotty release version
- Lots of new features for the theia & vscode integration
- A lot of efforts where made to align the node sever implemenation with the Java variant. Both server flavors are 
 now more ore less feature identicall
- Generic undo/redo support for node-based servers
- Support for running glsp servers directly in the browser via webworker (without the need for a backend process)
- and many, many more improvements

**Comment by PatrikN-work** [comment]
Sorry for reviving another thread. 

When I try this: https://github.com/eclipse-glsp/glsp-examples/tree/packing-demo i get this when installing the extension.

![Untitled Project](https://github.com/eclipse-glsp/glsp/assets/119344948/1002f540-9284-4a7b-a25a-2f558048f049)

When I open a file it is trying to load but nothing more.

Any help is appreciated

**Reply by rsoika** [reply]
This looks to me like your server is not running.

**Reply by PatrikN-work** [reply]
I guess, but i was under the impression that the setup that @tortmayr provided was complete and included that as well. Or am I missing something else?

It looks like the right things are generated in dist but I'm not sure what is picked up by vsce

**Reply by tortmayr** [reply]
Hi @PatrikN-work,
it seems like there was indeed something misconfigured which in the end lead to  an incomplete vsix (without the server bundles).
I have created a [PR](https://github.com/eclipse-glsp/glsp-examples/pull/211) that fixes this and adds packaging support to the `node-json-vsocde` template.

You can now create the vsix by calling `yarn package` from the root.
Could you please test it out and let me know if it works for you? Feel free to directly comment to the PR.

**Reply by PatrikN-work** [reply]
Hi @tortmayr ,

Just tried it and it seems to work fine now. Thanks!

/Patrik
